Mono.Debugging/Mono.Debugging.Client/DebuggerSession.cs

```diff
@@ -409,22 +409,20 @@ public void Run (DebuggerStartInfo startInfo, DebuggerSessionOptions options)
 				throw new ArgumentNullException (nameof (startInfo));
 			if (options == null)
 				throw new ArgumentNullException (nameof (options));
-			
-			lock (slock) {
-				this.options = options;
-				OnRunning ();
-				Dispatch (delegate {
-					try {
-						OnRun (startInfo);
-					} catch (Exception ex) {
-						// should handle exception before raising Exit event because HandleException may ignore exceptions in Exited state
-						var exceptionHandled = HandleException (ex);
-						ForceExit ();
-						if (!exceptionHandled)
-							throw;
-					}
-				});
-			}
+
+			Dispatch (delegate {
+				try {
+					this.options = options;
+					OnRunning ();
```

Comment:
I think this works. If the callbacks in the TargetStarted handler do end up calling other methods which require the lock, they will still have it, because it's the same thread owning the data. I can't think of a way to break this.